src/test/java/com/atlassian/db/replica/api/KeyBasedThrottledCacheTest.java

```diff
@@ -0,0 +1,110 @@
+package com.atlassian.db.replica.api;
+
+import org.junit.Test;
+
+import java.util.ConcurrentModificationException;
+import java.util.Optional;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class KeyBasedThrottledCacheTest {
+    private final ExecutorService executor = Executors.newFixedThreadPool(10);
+
+    @Test
+    public void staleWhileInvalidating() throws InterruptedException {
+        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(randomKey());
+        Long staleValue = 1L;
+        cache.get(() -> staleValue);
+
+        asyncCacheUpdateWithSlowSupplier(cache, 2L);
+
+        Optional<Long> actual = cache.get(() -> 3L);
+
+        assertThat(actual).hasValue(staleValue);
+    }
+
+
+    @Test
+    public void serveLatestValue() {
+        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(randomKey());
+        cache.get(() -> 1L);
+        cache.get(() -> 3L);
+
+        Long lastValue = 4L;
+        Optional<Long> actual = cache.get(() -> lastValue);
+
+        assertThat(actual).hasValue(lastValue);
+    }
+
+    @Test
+    public void noConcurrentInvalidation() throws InterruptedException {
+        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(randomKey());
+
+        asyncCacheUpdateWithSlowSupplier(cache, 2L);
+        cache.get(() -> {
+            throw new ConcurrentModificationException("Concurrent invalidation is not allowed");
+        });
+    }
+
+    @Test
+    public void cacheValuePerKey() throws InterruptedException {
+        String key = randomKey();
+        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(key);
+        KeyBasedThrottledCache cache2 = new KeyBasedThrottledCache(key);
+
+        cache.get(() -> 1L);
+
+        asyncCacheUpdateWithSlowSupplier(cache, 2L);
+        asyncCacheUpdateWithSlowSupplier(cache2, 3L);
+
+        Optional<Long> actual = cache2.get(this::anyValue);
```

Comment:
It's magical how the `cache` and `cache2` interact due to the hidden global state.

Reply:
`        cache.get(() -> 1L);
        asyncCacheUpdateWithSlowSupplier(cache, anyValue());

        cache2.get(this::anyValue);
        Optional<Long> actual = cache2.get(this::anyValue);

        assertThat(actual).hasValue(1L);`
        
        Is this easier to read?

Reply:
It's proving that when **cache1 blocks**, **cache2 is getting stale value from cache1**. 

_Detail: initial impl didn't need slow supplier for cache 2._

Reply:
The magic is an argument in the [keying discussion](https://github.com/atlassian-labs/db-replica/pull/68/files#r604088954). If keyed global state will remain, then the test is good as is. It accurately represents true behavior.